From: J. R. Okajima <hooanon05@yahoo.co.jp>
Date: Tue Jun 23 15:19:49 2015 +0900
Subject: aufs4.1 base patch
Origin: http://sourceforge.net/p/aufs/aufs4-standalone/ci/1a70d6a1618c28280e11783efed84d48e7cb15f2/tree/
Bug-Debian: https://bugs.debian.org/541828

Patch headers added by debian/patches/features/all/aufs4/gen-patch

aufs4.1 base patch

Index: linux-semplice-4.1.16-current/MAINTAINERS
===================================================================
--- linux-semplice-4.1.16-current.orig/MAINTAINERS
+++ linux-semplice-4.1.16-current/MAINTAINERS
@@ -1880,6 +1880,19 @@ F:	include/linux/audit.h
 F:	include/uapi/linux/audit.h
 F:	kernel/audit*
 
+AUFS (advanced multi layered unification filesystem) FILESYSTEM
+M:	"J. R. Okajima" <hooanon05g@gmail.com>
+L:	linux-unionfs@vger.kernel.org
+L:	aufs-users@lists.sourceforge.net (members only)
+W:	http://aufs.sourceforge.net
+T:	git://github.com/sfjro/aufs4-linux.git
+S:	Supported
+F:	Documentation/filesystems/aufs/
+F:	Documentation/ABI/testing/debugfs-aufs
+F:	Documentation/ABI/testing/sysfs-aufs
+F:	fs/aufs/
+F:	include/uapi/linux/aufs_type.h
+
 AUXILIARY DISPLAY DRIVERS
 M:	Miguel Ojeda Sandonis <miguel.ojeda.sandonis@gmail.com>
 W:	http://miguelojeda.es/auxdisplay.htm
Index: linux-semplice-4.1.16-current/drivers/block/loop.c
===================================================================
--- linux-semplice-4.1.16-current.orig/drivers/block/loop.c
+++ linux-semplice-4.1.16-current/drivers/block/loop.c
@@ -538,6 +538,24 @@ static inline int is_loop_device(struct
 	return i && S_ISBLK(i->i_mode) && MAJOR(i->i_rdev) == LOOP_MAJOR;
 }
 
+/*
+ * for AUFS
+ * no get/put for file.
+ */
+struct file *loop_backing_file(struct super_block *sb)
+{
+	struct file *ret;
+	struct loop_device *l;
+
+	ret = NULL;
+	if (MAJOR(sb->s_dev) == LOOP_MAJOR) {
+		l = sb->s_bdev->bd_disk->private_data;
+		ret = l->lo_backing_file;
+	}
+	return ret;
+}
+EXPORT_SYMBOL_GPL(loop_backing_file);
+
 /* loop sysfs attributes */
 
 static ssize_t loop_attr_show(struct device *dev, char *page,
Index: linux-semplice-4.1.16-current/fs/dcache.c
===================================================================
--- linux-semplice-4.1.16-current.orig/fs/dcache.c
+++ linux-semplice-4.1.16-current/fs/dcache.c
@@ -1167,7 +1167,7 @@ enum d_walk_ret {
  *
  * The @enter() and @finish() callbacks are called with d_lock held.
  */
-static void d_walk(struct dentry *parent, void *data,
+void d_walk(struct dentry *parent, void *data,
 		   enum d_walk_ret (*enter)(void *, struct dentry *),
 		   void (*finish)(void *))
 {
Index: linux-semplice-4.1.16-current/fs/read_write.c
===================================================================
--- linux-semplice-4.1.16-current.orig/fs/read_write.c
+++ linux-semplice-4.1.16-current/fs/read_write.c
@@ -494,6 +494,28 @@ ssize_t __vfs_write(struct file *file, c
 }
 EXPORT_SYMBOL(__vfs_write);
 
+vfs_readf_t vfs_readf(struct file *file)
+{
+	const struct file_operations *fop = file->f_op;
+
+	if (fop->read)
+		return fop->read;
+	if (fop->read_iter)
+		return new_sync_read;
+	return ERR_PTR(-ENOSYS);
+}
+
+vfs_writef_t vfs_writef(struct file *file)
+{
+	const struct file_operations *fop = file->f_op;
+
+	if (fop->write)
+		return fop->write;
+	if (fop->write_iter)
+		return new_sync_write;
+	return ERR_PTR(-ENOSYS);
+}
+
 ssize_t __kernel_write(struct file *file, const char *buf, size_t count, loff_t *pos)
 {
 	mm_segment_t old_fs;
Index: linux-semplice-4.1.16-current/fs/splice.c
===================================================================
--- linux-semplice-4.1.16-current.orig/fs/splice.c
+++ linux-semplice-4.1.16-current/fs/splice.c
@@ -1101,8 +1101,8 @@ EXPORT_SYMBOL(generic_splice_sendpage);
 /*
  * Attempt to initiate a splice from pipe to file.
  */
-static long do_splice_from(struct pipe_inode_info *pipe, struct file *out,
-			   loff_t *ppos, size_t len, unsigned int flags)
+long do_splice_from(struct pipe_inode_info *pipe, struct file *out,
+		    loff_t *ppos, size_t len, unsigned int flags)
 {
 	ssize_t (*splice_write)(struct pipe_inode_info *, struct file *,
 				loff_t *, size_t, unsigned int);
@@ -1118,9 +1118,9 @@ static long do_splice_from(struct pipe_i
 /*
  * Attempt to initiate a splice from a file to a pipe.
  */
-static long do_splice_to(struct file *in, loff_t *ppos,
-			 struct pipe_inode_info *pipe, size_t len,
-			 unsigned int flags)
+long do_splice_to(struct file *in, loff_t *ppos,
+		  struct pipe_inode_info *pipe, size_t len,
+		  unsigned int flags)
 {
 	ssize_t (*splice_read)(struct file *, loff_t *,
 			       struct pipe_inode_info *, size_t, unsigned int);
Index: linux-semplice-4.1.16-current/include/linux/file.h
===================================================================
--- linux-semplice-4.1.16-current.orig/include/linux/file.h
+++ linux-semplice-4.1.16-current/include/linux/file.h
@@ -19,6 +19,7 @@ struct dentry;
 struct path;
 extern struct file *alloc_file(struct path *, fmode_t mode,
 	const struct file_operations *fop);
+extern struct file *get_empty_filp(void);
 
 static inline void fput_light(struct file *file, int fput_needed)
 {
Index: linux-semplice-4.1.16-current/include/linux/fs.h
===================================================================
--- linux-semplice-4.1.16-current.orig/include/linux/fs.h
+++ linux-semplice-4.1.16-current/include/linux/fs.h
@@ -1663,6 +1663,12 @@ ssize_t rw_copy_check_uvector(int type,
 			      struct iovec *fast_pointer,
 			      struct iovec **ret_pointer);
 
+typedef ssize_t (*vfs_readf_t)(struct file *, char __user *, size_t, loff_t *);
+typedef ssize_t (*vfs_writef_t)(struct file *, const char __user *, size_t,
+				loff_t *);
+vfs_readf_t vfs_readf(struct file *file);
+vfs_writef_t vfs_writef(struct file *file);
+
 extern ssize_t __vfs_read(struct file *, char __user *, size_t, loff_t *);
 extern ssize_t __vfs_write(struct file *, const char __user *, size_t, loff_t *);
 extern ssize_t vfs_read(struct file *, char __user *, size_t, loff_t *);
Index: linux-semplice-4.1.16-current/include/linux/splice.h
===================================================================
--- linux-semplice-4.1.16-current.orig/include/linux/splice.h
+++ linux-semplice-4.1.16-current/include/linux/splice.h
@@ -83,4 +83,10 @@ extern void splice_shrink_spd(struct spl
 extern void spd_release_page(struct splice_pipe_desc *, unsigned int);
 
 extern const struct pipe_buf_operations page_cache_pipe_buf_ops;
+
+extern long do_splice_from(struct pipe_inode_info *pipe, struct file *out,
+			   loff_t *ppos, size_t len, unsigned int flags);
+extern long do_splice_to(struct file *in, loff_t *ppos,
+			 struct pipe_inode_info *pipe, size_t len,
+			 unsigned int flags);
 #endif
